### PR TITLE
fs: implement unix OpenOptionsExt for OpenOptions

### PR DESCRIPTION
**Pull request by chovine**
## Motivation

Implement `std::os::unix::fs::OpenOptionsExt` for `tokio::fs::OpenOptions` in order to avoid to have to create `tokio::fs::OpenOptions` from `std::fs::OpenOptions` when using OpenOptionsExt.

## Solution

- Added an `as_inner_mut` method to `tokio::fs::OpenOptions`, visible only to the `tokio::fs` module.
- Added the trivial implementation of OpenOptionsExt to `open_options_ext.rs`, using the underlying `std::fs::OpenOptions` of `tokio::fs::OpenOptions`.
- Mirrored the definition of `std::os::unix::fs::OpenOptionsExt` in `tokio::fs::os::unix`

### CODE REVIEW

File: tokio/src/fs/open_options.rs [hunk 1]
```diff
@@ -382,6 +382,13 @@ impl OpenOptions {
         let std = asyncify(move || opts.open(path)).await?;
         Ok(File::from_std(std))
     }
+
+    /// Returns an immutable reference to the the underlying std::fs::OpenOptions
+    #[doc(hidden)]
```
Comment (Darksonn) [inline tokio/src/fs/open_options.rs:387]: The `#[doc(hidden)]` attribute is not necessary on private methods.

File: tokio/src/fs/os/unix/mod.rs [hunk 2]
```diff
@@ -2,3 +2,6 @@
 
 mod symlink;
 pub use self::symlink::symlink;
+
+mod open_options_ext;
+pub use self::open_options_ext::OpenOptionsExt;
```
[the remark this answers is not in the document]
Comment (chovine) [reply tokio/src/fs/os/unix/mod.rs:7]: The idea was to expose it in a way similar to `std::os::unix::fs::OpenOptionsExt` (`tokio::fs::unix::OpenOptionsExt` in this case) as `symlink` and `OpenOptionsExt` are at the same level in the standard library.

File: tokio/src/fs/open_options.rs (hunk 1 above)
Comment (chovine) [reply tokio/src/fs/open_options.rs:387]: will remove, thanks

**Comment by sfackler** [comment]
I would recommend against directly implementing std's trait. Tokio's build will break whenever new methods are added to it.

File: tokio/src/fs/os/unix/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Darksonn) [reply tokio/src/fs/os/unix/mod.rs:7]: Well `synlink` is not a reexport — it's a special async version, whereas the `OpenOptionsExt` trait is the same in both cases.

**Comment by Darksonn** [comment]
@sfackler But that would be a breaking change in std, no?

**Comment by sfackler** [comment]
We generally assume that those traits have not been implemented externally. They should probably be sealed, but that was invented after the traits were added. custom_flags was added in 1.10, for example.

File: tokio/src/fs/os/unix/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chovine) [reply tokio/src/fs/os/unix/mod.rs:7]: Fair point, I'll remove the reexport.

**Comment by Darksonn** [comment]
I see. I guess we should add our own duplicate of the trait then.

**Comment by chovine** [comment]
> I see. I guess we should add our own duplicate of the trait then.

I can redefine `OpenOptionExt` along with the trait implementation. Should this be a separate issue?

**Comment by Darksonn** [comment]
You can just do it here.

File: tokio/src/fs/os/unix/open_options_ext.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> io::Result<()> {
+    ///     let mut options = OpenOptions::new();
+    ///     options.mode(0o644); // Give read/write for owner and read for others.
+    ///     let file = options.open("foo.txt").await?;
+    ///
+    ///     Ok(())
+    ///
+    /// }
```
Comment (Darksonn) [inline tokio/src/fs/os/unix/open_options_ext.rs:35]:
```suggestion
    ///     Ok(())
    /// }
```

File: tokio/src/fs/open_options.rs [hunk 4]
```diff
@@ -382,6 +382,12 @@ impl OpenOptions {
         let std = asyncify(move || opts.open(path)).await?;
         Ok(File::from_std(std))
     }
+
+    /// Returns an immutable reference to the the underlying std::fs::OpenOptions
```
Comment (Darksonn) [inline tokio/src/fs/open_options.rs:386]:
```suggestion
    /// Returns a mutable reference to the the underlying std::fs::OpenOptions
```

**Comment by chovine** [comment]
Both remarks included, thanks for your review.

File: tokio/src/fs/os/unix/open_options_ext.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
+use crate::fs::open_options::OpenOptions;
+use std::os::unix::fs::OpenOptionsExt as StdOpenOptionsExt;
+
+/// Unix-specific extensions to [`fs::OpenOptions`].
+///
+/// This mirrors the definition of [`std::os::unix::fs::OpenOptionsExt`].
+///
+///
+/// [`fs::OpenOptions`]: ../../struct.OpenOptions.html
```
Comment (Darksonn) [inline tokio/src/fs/os/unix/open_options_ext.rs:9]: This should be a intra-rustdoc link.
```suggestion
/// [`fs::OpenOptions`]: crate::fs::OpenOptions
```

Comment (chovine) [reply tokio/src/fs/os/unix/open_options_ext.rs]: Fixed, thanks!

**Comment by Darksonn** [comment]
CI is messed up. Can you please push an empty commit to fix it?

**Comment by Darksonn** [comment]
This is a bit ridiculous. You don't need to push more empty commits, I'll do something about it if it doesn't work this time around.

**Comment by chovine** [comment]
> This is a bit ridiculous. You don't need to push more empty commits, I'll do something about it if it doesn't work this time around.

I assumed the timeouts were due to the CI runner being under heavy load. Feel free to fix if it does not work this time

**Comment by Darksonn** [comment]
No, there's some issue with the windows tests in #2499, but we aren't many windows users, and the few of us that do use windows have not been able to reproduce it on their own machines.

**Pull request merged**